include/quicr/message_buffer.h

```diff
@@ -106,87 +104,145 @@ class MessageBuffer
   MessageBuffer& operator=(MessageBuffer&& other) = default;
 
-  template<typename Uint_t>
-  friend MessageBuffer& operator<<(MessageBuffer& msg, Uint_t val);
-  template<typename Uint_t>
-  friend MessageBuffer& operator>>(MessageBuffer& msg, Uint_t& val);
+  /**
+   * @brief A fancy operator for push, writes a byte on the end of the buffer.
+   * @param buffer The MessageBuffer to push to.
+   * @param value The byte to push.
+   * @returns The MessageBuffer that was written to.
+   */
+  friend inline MessageBuffer& operator<<(MessageBuffer& buffer, uint8_t value)
+  {
+    buffer._buffer.push_back(value);
+    return buffer;
+  }
 
-private:
-  std::vector<uint8_t> _buffer;
-};
+  /**
+   * @brief Writes QUICR integral types to the buffer in NBO.
+   * @tparam T A type satisfying quicr::is_integral.
+   * @param msg The MessageBuffer to write to.
+   * @param value The message to be written.
+   * @returns The MessageBuffer that was written to.
+   */
+  template<typename T, typename = std::enable_if_t<quicr::is_integral_v<T>, T>>
+  friend inline MessageBuffer& operator<<(MessageBuffer& msg, T value)
+  {
+    value = swap_bytes(value);
+    uint8_t* val_ptr = reinterpret_cast<uint8_t*>(&value);
 
-MessageBuffer&
-operator<<(MessageBuffer& msg, const uintVar_t& val);
-MessageBuffer&
-operator>>(MessageBuffer& msg, uintVar_t& val);
+    const auto length = msg._buffer.size();
+    msg._buffer.resize(length + sizeof(T));
+    std::memcpy(msg._buffer.data() + length, val_ptr, sizeof(T));
 
-MessageBuffer&
-operator<<(MessageBuffer& msg, const std::vector<uint8_t>& val);
-MessageBuffer&
-operator<<(MessageBuffer& msg, std::vector<uint8_t>&& val);
-MessageBuffer&
-operator>>(MessageBuffer& msg, std::vector<uint8_t>& val);
+    return msg;
+  }
 
-namespace {
-// clang-format off
-constexpr uint16_t
-swap_bytes(uint16_t value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+  /**
+   * @brief Overload for Namespace integral type to write in NBO.
+   * @param msg The MessageBuffer to write to.
+   * @param value The message to be written.
+   * @returns The MessageBuffer that was written to.
+   */
+  friend MessageBuffer& operator<<(MessageBuffer& msg, quicr::Namespace value)
+  {
+    if constexpr (std::endian::native == std::endian::big)
+      return msg << value.name() << value.length();
 
-  return ((value >> 8) & 0x00ff) | ((value << 8) & 0xff00);
-}
+    return msg << value.length() << value.name();
+  }
 
-constexpr uint32_t
-swap_bytes(uint32_t value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+  /**
+   * @brief A fancy operator for pop, reads a byte off the buffer.
+   * @param buffer The MessageBuffer to read from.
+   * @param value The value to read into.
+   * @returns The MessageBuffer that was read from.
+   */
+  friend inline MessageBuffer& operator>>(MessageBuffer& msg, uint8_t& value)
+  {
+    if (msg.empty()) {
```

Comment:
While this addresses an empty buffer, since some calls to front() might demand more data than the buffer has, why not just have an exception thrown from front()?  It would reduce how many places one throws an exception.